crates/node/src/storage/rocksdb.rs

```diff
@@ -1817,113 +1816,130 @@ impl DB for RocksDB {
+        for st in StoreType::iter() {
+            if st.is_stored_every_block() {
+                let key_prefix = if st.is_stored_every_block() {
+                    tree_key_prefix_with_height(st, height)
+                } else {
+                    tree_key_prefix_with_epoch(st, epoch)
+                };
+                let root_key =
+                    format!("{key_prefix}/{MERKLE_TREE_ROOT_KEY_SEGMENT}");
+                self.add_value_to_batch(
+                    block_cf,
+                    root_key,
+                    merkle_tree_stores.root(st),
+                    &mut batch,
+                );
+                let store_key =
+                    format!("{key_prefix}/{MERKLE_TREE_STORE_KEY_SEGMENT}");
+                self.add_value_bytes_to_batch(
+                    block_cf,
+                    store_key,
+                    merkle_tree_stores.store(st).encode(),
+                    &mut batch,
+                );
             }
         }
-        Ok(db_visitor.take_batch())
+        self.exec_batch(batch)
     }
 }
 
 /// A struct that can visit a set of updates,
 /// registering them all in the batch
-pub struct RocksDBUpdateVisitor<'db> {
-    db: &'db RocksDB,
+#[derive(Default)]
+pub struct RocksDBUpdateVisitor {
     batch: RocksDBWriteBatch,
 }
 
-impl<'db> RocksDBUpdateVisitor<'db> {
-    pub fn new(db: &'db RocksDB) -> Self {
-        Self {
-            db,
-            batch: Default::default(),
-        }
-    }
+impl DBUpdateVisitor for RocksDBUpdateVisitor {
+    type DB = RocksDB;
 
-    pub fn take_batch(self) -> RocksDBWriteBatch {
-        self.batch
-    }
-}
-
-impl<'db> DBUpdateVisitor for RocksDBUpdateVisitor<'db> {
-    fn read(&self, key: &Key, cf: &DbColFam) -> Option<Vec<u8>> {
+    fn read(&self, db: &Self::DB, key: &Key, cf: &DbColFam) -> Option<Vec<u8>> {
         match cf {
-            DbColFam::SUBSPACE => self
-                .db
+            DbColFam::SUBSPACE => db
                 .read_subspace_val(key)
                 .expect("Failed to read from storage"),
             _ => {
                 let cf_str = cf.to_str();
-                let cf = self
-                    .db
+                let cf = db
                     .get_column_family(cf_str)
                     .expect("Failed to read column family from storage");
-                self.db
-                    .read_value_bytes(cf, key.to_string())
+                db.read_value_bytes(cf, key.to_string())
                     .expect("Failed to get key from storage")
             }
         }
     }
 
-    fn write(&mut self, key: &Key, cf: &DbColFam, value: impl AsRef<[u8]>) {
-        self.db
-            .overwrite_entry(&mut self.batch, cf, key, value)
+    fn write(
+        &mut self,
+        db: &Self::DB,
+        key: &Key,
+        cf: &DbColFam,
+        value: impl AsRef<[u8]>,
+    ) {
+        db.overwrite_entry(&mut self.batch, cf, key, value)
```

Comment:
It seems that it could be written to the diff storage even if the key should be for no diff.
Can we use `db.batch_write_subspace_val()` for the subspace keys in `overwrite_entry`?

Reply:
good question - unfortunately we cannot pass a column family to `batch_write_subspace_val` as it's only for subspaces, but I think we can extend `overwrite_entry` to pass through a flag to decide. We also need to do the same for `fn delete` below this one (currently it's hard-coded to `persist_diffs = true`)

Reply:
added [d59a87b](https://github.com/anoma/namada/pull/4135/commits/d59a87b1341a1e98b9437d1b703b3ebac3983dba)